docs/docs/proposals/007-grafanalibrarypanel-crd.md

```diff
@@ -0,0 +1,238 @@
+custom resource. It can thus be provided by the CR owner. By setting the UID to a stable value
+as opposed to letting Grafana autogenerate it, it's possible to provision both Library Panels
+and Dashboards that reference them as CRs.
+
+#### Dynamic referencing
+
+Similar to how Datasources can be dynamically referenced via the
+`.spec.datasources` field, we can extend the GrafanaDashboard CR to similarly
+allow mapping library panels via a new `.spec.libraryPanels` field structured
+very similarly. The CR author can specify a replacement string and a reference
+to the library panel deployed itself as a CR.
+
+When reconciling the GrafanaDashboard, the controller will fetch the dependent
+GrafanaLibraryPanel CRs, look up their UIDs, and then perform the string
+replacement.
+
+```yaml
+apiVersion: grafana.integreatly.org/v1beta1
+kind: GrafanaDashboard
+metadata:
+  name: grpc-overview
+spec:
+  libraryPanels:
+    - inputName: "LP_GRPC_SUCCESS_RATE"
+      libraryPanelRef: "grpc-server-success-rate"
+  allowCrossNamespaceImport: true
+  contentCacheDuration: 1m0s
+  folderRef: folder-ref
+  instanceSelector:
+    matchLabels:
+      env: dev
+      region: us-central1
+  json: >
+    {
+      "__inputs": [
+        {
+          "name": "DS_PROMETHEUS",
+          "label": "prometheus",
+          "description": "",
+          "type": "datasource",
+          "pluginId": "prometheus",
+          "pluginName": "Prometheus"
+        }
+      ],
+      "__elements": {},
+      "__requires": [
+        {
+          "type": "grafana",
+          "id": "grafana",
+          "name": "Grafana",
+          "version": "9.1.6"
+        },
+        {
+          "type": "datasource",
+          "id": "prometheus",
+          "name": "Prometheus",
+          "version": "1.0.0"
+        },
+        {
+          "type": "panel",
+          "id": "timeseries",
+          "name": "Time series",
+          "version": ""
+        }
+      ],
+      "annotations": {
+        "list": [
+          {
+            "builtIn": 1,
+            "datasource": {
+              "type": "grafana",
+              "uid": "-- Grafana --"
+            },
+            "enable": true,
+            "hide": true,
+            "iconColor": "rgba(0, 211, 255, 1)",
+            "name": "Annotations & Alerts",
+            "target": {
+              "limit": 100,
+              "matchAny": false,
+              "tags": [],
+              "type": "dashboard"
+            },
+            "type": "dashboard"
+          }
+        ]
+      },
+      "editable": true,
+      "fiscalYearStartMonth": 0,
+      "graphTooltip": 1,
+      "id": null,
+      "links": [],
+      "liveNow": false,
+      "panels": [
+        {
+          "gridPos": {
+            "h": 8,
+            "w": 12,
+            "x": 0,
+            "y": 0
+          },
+          "id": 2,
+          "libraryPanel": {
+            "uid": "${LP_GRPC_SERVER_SUCCESS_RATE}",
+            "name": "gRPC Server Success Rate"
```

Comment:
tbh I am not sure what this `.name` field is even used for. It could possibly also be dynamically injected, and we could have this just be `"libraryPanel": "${LP_GRPC_SERVER_SUCCESS_RATE}"`, but that might be a bit too surprising of an interface.

Reply:
I tested importing this with an arbitrary value, and it seems to work, so I guess we can just ignore that field